File: src/tortoisebot_waypoints/tortoisebot_action_server.py
#! /usr/bin/env python3
import rospy
import time
import actionlib

from tortoisebot_waypoints.msg import WaypointActionFeedback, WaypointActionResult, WaypointActionAction
from geometry_msgs.msg import Twist
from std_msgs.msg import Empty
from geometry_msgs.msg import Twist, Point
from nav_msgs.msg import Odometry
from tf import transformations
import math

def normalize_angle(angle):
    # Normalize angle to be within [-pi, pi)
    normalized_angle = (angle + math.pi) % (2 * math.pi) - math.pi
    return normalized_angle

class WaypointActionClass(object):

    # create messages that are used to publish feedback/result
    _feedback = WaypointActionFeedback()
    _result = WaypointActionResult()

    # topics
    _pub_cmd_vel = None
    _sub_odom = None

    # go to point vars
    # robot state variables
    _position = Point()
    _yaw = 0
    # machine state
    _state = 'idle'
    # goal
    _des_pos = Point()
    _des_yaw = 0.0
    # parameters
    _yaw_precision = math.pi / 90 # +/- 2 degree allowed
    _dist_precision = 0.05
    _max_angular_vel_error_yaw = 10*_yaw_precision  # [rad] 20
    _max_speed_error_pos = _dist_precision  # [m]
    _max_angular_vel = 0.65  # [rad/s]
    _max_speed = 0.1  # [m/s]

    def __init__(self):
        # creates the action server
        self._as = actionlib.SimpleActionServer("tortoisebot_as", WaypointActionAction, self.goal_callback, False)
        self._as.start()

        # define a loop rate
        self._rate = rospy.Rate(25)

        # topics
        self._pub_cmd_vel = rospy.Publisher('/cmd_vel', Twist, queue_size=1)
        rospy.wait_for_message('/odom', Odometry)
        self._sub_odom = rospy.Subscriber('/odom', Odometry, self._clbk_odom)
        rospy.loginfo("Action server started")

    def _clbk_odom(self, msg):
        # position
        self._position = msg.pose.pose.position

        # yaw
        quaternion = (
            msg.pose.pose.orientation.x,
            msg.pose.pose.orientation.y,
            msg.pose.pose.orientation.z,
            msg.pose.pose.orientation.w)
        euler = transformations.euler_from_quaternion(quaternion)
        self._yaw = euler[2]

    def goal_callback(self, goal):
        rospy.loginfo("goal %s received" % str(goal))

        # helper variables
        success = True

        # define desired position
        self._des_pos = goal.position
        self._des_yaw = goal.yaw

        # perform task
        while success:
            # update vars
            err_pos = math.sqrt(pow(self._des_pos.y - self._position.y, 2) + pow(self._des_pos.x - self._position.x, 2))
            if err_pos > self._dist_precision:
                desired_yaw = math.atan2(self._des_pos.y - self._position.y, self._des_pos.x - self._position.x)
            else:
                desired_yaw = self._des_yaw
            err_yaw = normalize_angle(desired_yaw - self._yaw)
            rospy.loginfo("Current Yaw: %s" % str(self._yaw))
            rospy.loginfo("Desired Yaw: %s" % str(desired_yaw))
            rospy.loginfo("Error Yaw: %s" % str(err_yaw))
            # logic goes here
            if err_pos <= self._dist_precision and math.fabs(err_yaw) <= self._yaw_precision:
                break
            elif self._as.is_preempt_requested():
                # cancelled
                rospy.loginfo("The goal has been cancelled/preempted")
                self._as.set_preempted()
                success = False
            elif math.fabs(err_yaw) > self._yaw_precision:
                # fix yaw
                rospy.loginfo("fix yaw")
                self._state = 'fix yaw'
                twist_msg = Twist()
                twist_msg.angular.z = max(min(err_yaw*self._max_angular_vel/self._max_angular_vel_error_yaw, self._max_angular_vel), -self._max_angular_vel)
                rospy.loginfo("publishing w_z: %s" % str(twist_msg.angular.z))
                self._pub_cmd_vel.publish(twist_msg)
            else:
                # go to point
                rospy.loginfo("go to point")
                self._state = 'go to point'
                twist_msg = Twist()
                twist_msg.linear.x = max(min(err_pos*self._max_speed/self._max_speed_error_pos, self._max_speed), -self._max_speed)
                twist_msg.angular.z = max(min(err_yaw*self._max_angular_vel/self._max_angular_vel_error_yaw, self._max_angular_vel), -self._max_angular_vel)
                self._pub_cmd_vel.publish(twist_msg)

            # send feedback
            self._feedback.position = self._position
            self._feedback.yaw = self._yaw
            self._feedback.state = self._state
            self._as.publish_feedback(self._feedback)

            # loop rate
            self._rate.sleep()

        # stop
        twist_msg = Twist()
        twist_msg.linear.x = 0
        twist_msg.angular.z = 0
        self._pub_cmd_vel.publish(twist_msg)

        # return success
        if success:
            self._result.success = True
            self._as.set_succeeded(self._result)

if __name__ == '__main__':
    rospy.init_node('tortoisebot_as')
    WaypointActionClass()
    rospy.spin()

File: src/tortoisebot_action_server.cpp
#include "tortoisebot_waypoints/tortoisebot_action_server.hpp"

#include "rclcpp/logging.hpp"
#include "rclcpp/node.hpp"
#include "rclcpp_action/rclcpp_action.hpp"

#include "geometry_msgs/msg/quaternion.hpp"
#include "tf2/LinearMath/Matrix3x3.h"
#include "tf2/LinearMath/Quaternion.h"

#include <algorithm>
#include <cmath>
#include <functional>
#include <mutex>

using namespace TortoisebotWaypoints;

// Shift angle to [-pi, pi)
double normalize_angle(double angle) {
  constexpr double pi{3.1416};
  angle = std::fmod(angle + pi, 2 * pi);

  if (angle < 0) {
    angle += 2 * pi;
  }

  return angle - pi;
}

TortoisebotActionServer::TortoisebotActionServer()
    : Node{kNodeName}, cmd_vel_pub_{this->create_publisher<Twist>(
                           kCmdVelTopicName, 10)},
      odom_sub_{this->create_subscription<Odometry>(
          kOdomTopicName, 1,
          std::bind(&TortoisebotActionServer::odom_callback, this,
                    std::placeholders::_1))},
      action_server_{rclcpp_action::create_server<WaypointAction>(
          this, kActionName,
          std::bind(&TortoisebotActionServer::handle_goal, this,
                    std::placeholders::_1, std::placeholders::_2),
          std::bind(&TortoisebotActionServer::handle_cancel, this,
                    std::placeholders::_1),
          std::bind(&TortoisebotActionServer::handle_accepted, this,
                    std::placeholders::_1))} {
  RCLCPP_INFO(this->get_logger(), "Started %s action server.", kActionName);
}

TortoisebotActionServer::~TortoisebotActionServer() {
  cleanup_goal_threads();
}

void TortoisebotActionServer::odom_callback(
    const std::shared_ptr<const Odometry> msg) {
  tf2::Quaternion q{msg->pose.pose.orientation.x, msg->pose.pose.orientation.y,
                    msg->pose.pose.orientation.z, msg->pose.pose.orientation.w};
  double yaw{};
  double pitch{};
  double roll{};
  tf2::Matrix3x3(q).getEulerYPR(yaw, pitch, roll);

  std::lock_guard<std::mutex> lock{state_mutex_};
  position_ = msg->pose.pose.position;
  yaw_ = yaw;
}

rclcpp_action::GoalResponse TortoisebotActionServer::handle_goal(
    const rclcpp_action::GoalUUID & /* uuid */,
    std::shared_ptr<const WaypointAction::Goal> goal) {

  RCLCPP_INFO(this->get_logger(), "Goal received: (x=%f, y=%f, theta=%f).",
              goal->position.x, goal->position.y, goal->yaw);
  return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
}

rclcpp_action::CancelResponse TortoisebotActionServer::handle_cancel(
    const std::shared_ptr<GoalHandleWaypointAction> /* goal_handle */) {
  RCLCPP_INFO(this->get_logger(), "Cancelling goal...");
  return rclcpp_action::CancelResponse::ACCEPT;
}

void TortoisebotActionServer::handle_accepted(
    const std::shared_ptr<GoalHandleWaypointAction> goal_handle) {
  if (current_goal_handle_ && current_goal_handle_->is_active()) {
    RCLCPP_INFO(this->get_logger(), "Preempting current goal...");
    current_goal_handle_->abort(std::make_shared<WaypointAction::Result>());
  }

  current_goal_handle_ = goal_handle;

  cleanup_goal_threads();
  goal_thread_ = std::thread{
      std::bind(&TortoisebotActionServer::execute, this, std::placeholders::_1),
      goal_handle};
}

void TortoisebotActionServer::execute(
    const std::shared_ptr<GoalHandleWaypointAction> goal_handle) {

  // Extract goal
  auto goal{goal_handle->get_goal()};
  auto feedback{std::make_shared<WaypointAction::Feedback>()};
  auto result{std::make_shared<WaypointAction::Result>()};

  double err_pos{}, desired_yaw{}, err_yaw{};
  bool keep_looping{true};

  while (rclcpp::ok() && keep_looping) {
    // Compute errors
    {
      std::lock_guard<std::mutex> lock{state_mutex_};
      err_pos = std::sqrt(std::pow(goal->position.x - position_.x, 2) +
                          std::pow(goal->position.y - position_.y, 2));
      desired_yaw = err_pos > dist_precision_
                        ? std::atan2(goal->position.y - position_.y,
                                     goal->position.x - position_.x)
                        : goal->yaw;
      err_yaw = normalize_angle(desired_yaw - yaw_);
    }

    // Main loop
    if (err_pos <= dist_precision_ && std::fabs(err_yaw) <= yaw_precision_) {
      RCLCPP_INFO(this->get_logger(), "Reached goal: (x=%f, y=%f, theta=%f).",
                  goal->position.x, goal->position.y, goal->yaw);
      if (goal_handle->is_active()) {
        result->success = true;
        goal_handle->succeed(result);
      }
      feedback->state = "idle";
      keep_looping = false;
    } else if (goal_handle->is_canceling()) {
      RCLCPP_INFO(this->get_logger(), "Cancelled goal: (x=%f, y=%f, theta=%f).",
                  goal->position.x, goal->position.y, goal->yaw);
      result->success = false;
      goal_handle->canceled(result);
      feedback->state = "idle";
      keep_looping = false;
    } else if (!goal_handle->is_active()) {
      RCLCPP_INFO(this->get_logger(), "Aborted goal: (x=%f, y=%f, theta=%f).",
                  goal->position.x, goal->position.y, goal->yaw);
      result->success = false;
      feedback->state = "idle";
      keep_looping = false;
    } else if (std::fabs(err_yaw) > yaw_precision_) {
      // Fix yaw
      feedback->state = "fix yaw";
      Twist cmd_vel_msg{};
      cmd_vel_msg.angular.z = std::max(std::min(err_yaw*max_angular_vel_/max_angular_vel_error_yaw_, max_angular_vel_), -max_angular_vel_);
      cmd_vel_pub_->publish(cmd_vel_msg);
    } else {
      // Move towards point
      feedback->state = "go to point";
      Twist cmd_vel_msg{};
      cmd_vel_msg.angular.z = std::max(std::min(err_yaw*max_angular_vel_/max_angular_vel_error_yaw_, max_angular_vel_), -max_angular_vel_);
      cmd_vel_msg.linear.x = std::max(std::min(err_pos*max_speed_/max_speed_error_pos_, max_speed_), -max_speed_);
      cmd_vel_pub_->publish(cmd_vel_msg);
    }

    feedback->position = position_;
    feedback->yaw = yaw_;
    goal_handle->publish_feedback(feedback);

    rate_.sleep();
  }

  // Stop robot
  Twist stop_msg{};
  cmd_vel_pub_->publish(stop_msg);
}

void TortoisebotActionServer::cleanup_goal_threads() {
  if (goal_thread_.joinable()) {
    goal_thread_.join();
  }
}
